File: solutions/c/largest-series-product/1/largest_series_product.c
#include "largest_series_product.h"

#include <string.h>
#include <ctype.h>

int64_t largest_series_product(char *digits, size_t span){
    size_t len = strlen(digits);
    if(span > len || (len == 0 && span >0)) {
        return -1;
    }
    int64_t max=0;
    for(size_t i=0; i<=len-span; i++){
        int64_t current = 1;
        for(size_t j=i; j<i+span; j++){
            if(!isdigit(digits[j]))
                return -1;
            current *= (digits[j] - '0');
        }
        max = current>max ? current : max;
    }
    return max;    
}

File: solutions/dart/largest-series-product/1/lib/largest_series_product.dart
import 'dart:math';

class LargestSeriesProduct {

 

  int largestProduct(String digitString,  int span){
    if (span<0)
      throw ArgumentError('span must not be negative');
    else if (span>digitString.length)
      throw ArgumentError('span must be smaller than string length'); 
    else if (span==0)
      return 1;
    
    List<int> digits =
     RegExp(r"\d")
       .allMatches(digitString)
       .map((m)=>m.group(0))
       .nonNulls
       .map((d) => int.tryParse(d))
       .nonNulls.toList();
    if(digits.length != digitString.length)
      throw ArgumentError('digits input must only contain digits');

    int largest=0;
    for(var i=0; i<digits.length - span+1; i++){
      int product = digits.sublist(i,i+span)
        .reduce((a,b)=>a*b);
      if(product>largest) 
        largest = product;
    }
    return largest;
  }
  
}
